src/hotspot/share/prims/unsafe.cpp

```diff
@@ -218,44 +218,25 @@ class MemoryAccess : StackObj {
   }
 
   T get() {
-    if (_obj == NULL) {
-      GuardUnsafeAccess guard(_thread);
-      T ret = RawAccess<>::load(addr());
-      return normalize_for_read(ret);
-    } else {
-      T ret = HeapAccess<>::load_at(_obj, _offset);
-      return normalize_for_read(ret);
-    }
+    GuardUnsafeAccess guard(_thread);
+    return normalize_for_read(*addr());
   }
 
   void put(T x) {
-    if (_obj == NULL) {
-      GuardUnsafeAccess guard(_thread);
-      RawAccess<>::store(addr(), normalize_for_write(x));
-    } else {
-      HeapAccess<>::store_at(_obj, _offset, normalize_for_write(x));
-    }
+    GuardUnsafeAccess guard(_thread);
+    *addr() = normalize_for_write(x);
   }
 
 
   T get_volatile() {
-    if (_obj == NULL) {
-      GuardUnsafeAccess guard(_thread);
-      volatile T ret = RawAccess<MO_SEQ_CST>::load(addr());
-      return normalize_for_read(ret);
-    } else {
-      T ret = HeapAccess<MO_SEQ_CST>::load_at(_obj, _offset);
-      return normalize_for_read(ret);
-    }
+    GuardUnsafeAccess guard(_thread);
+    volatile T ret = Atomic::load_acquire(addr());
```

Comment:
Is MO_SEQ_CST to load_acquire mapping correct?

Reply:
No it is not.

Reply:
> Is MO_SEQ_CST to load_acquire mapping correct?

We should probably retain RawAccess<MO_SEQ_CST> here to get the right semantics.

Reply:
I'll restore the volatile accesses to use RawAccess<MO_SEQ_CST> for now, but maybe it would be better in the long run to have simpler Atomic::load<MO_SEQ_CST> accessors?